OCPBUGS-18132: Block 4.14 upgrades with admin ack

gnufied opened the pull request:
Fixes https://issues.redhat.com/browse/OCPBUGS-18132

gnufied commented:
/jira refresh

gnufied commented:
/jira refresh

jsafrane commented on pkg/operator/vspherecontroller/checks/check_error.go:
```diff
@@ -163,11 +191,74 @@ func CheckClusterStatus(result ClusterCheckResult, apiDependencies KubeAPIInterf
+		return ClusterCheckDegrade, MakeClusterDegradedError(CheckStatusOpenshiftAPIError, reason)
+	}
+
+	klog.Infof("Checking for intree plugin use")
```
nit: `V(2)`

jsafrane commented on pkg/operator/vspherecontroller/checks/check_error.go:
```diff
@@ -163,11 +191,74 @@ func CheckClusterStatus(result ClusterCheckResult, apiDependencies KubeAPIInterf
+
+	for _, pv := range pvs {
+		if pv.Spec.VsphereVolume != nil {
+			usingvSphereVolumes = true
```
Would it be useful to log what PV has in-tree volume? In case an user argues they moved all PVs to CSI.

gnufied replied:
added logs

jsafrane replied:
Err.. I meant log the PV names.

gnufied replied:
hmm, I wonder if that will be a bit much. In larger clusters this could be a big string...

jsafrane commented on pkg/operator/vspherecontroller/checks/check_error.go:
```diff
@@ -163,11 +191,74 @@ func CheckClusterStatus(result ClusterCheckResult, apiDependencies KubeAPIInterf
+		volumesInUse := node.Status.VolumesInUse
+		for _, volume := range volumesInUse {
+			if strings.HasPrefix(string(volume), inTreePluginName) {
+				return true
```
Would it be useful to log the inline volume? In case an user argues they moved all Pods to CSI.

gnufied replied:
added logs.

jsafrane commented on pkg/operator/vspherecontroller/vspherecontroller.go:
```diff
@@ -349,6 +364,11 @@ func (c *VSphereController) blockUpgradeOrDegradeCluster(
+	case checks.ClusterCheckUpgradesBlockedViaAdminAck:
+		clusterCondition = "upgrade_blocked_admin_ack"
+		utils.InstallErrorMetric.WithLabelValues(string(result.CheckStatus), clusterCondition).Set(1)
+		updateError := c.addRequiresAdminAck(ctx, result, c.name)
+		return updateError, false, false
```
Where is the admin-ack _removed_ when user moves all PVs to CSI?

gnufied replied:
Added some code for it for removing the admin-ack keys when conditions are right. But this needs to be carefully reviewed. See - https://github.com/openshift/vmware-vsphere-csi-driver-operator/pull/169/files#diff-bd7b198fe282f7e27754079508639882be175cc903f563fb69f2c89fc6875a2aR319

jsafrane commented on pkg/operator/vspherecontroller/vspherecontroller.go:
```diff
@@ -494,6 +515,23 @@ func (c *VSphereController) updateConditions(
+		return fmt.Errorf("failed to get admin-gate configmap: %v", err)
+	}
+
+	klog.Infof("Updating admin-gates")
```
nit: `V(2)`

gnufied commented on pkg/operator/vspherecontroller/vsphere_environment_checker_test.go:
```diff
@@ -31,14 +32,28 @@ func TestEnvironmentCheck(t *testing.T) {
+			initialObjects:         []runtime.Object{testlib.GetConfigMap(), testlib.GetSecret()},
+			configObjects:          runtime.Object(testlib.GetInfraObject()),
+			clusterCSIDriverObject: testlib.MakeFakeDriverInstance(),
+			vcenterVersion:         "7.0.2",
```
Added some tests for 8 series. But I can add more.

jsafrane commented on pkg/operator/vspherecontroller/checks/check_patched_vcenter.go:
```diff
@@ -0,0 +1,85 @@
+			return true, fmt.Errorf("error converting build number %s to integer", build)
+		}
+
+		if buildNumber >= minBuildNumber8Series {
```
Just an idea: we know build number of the last wrong version. Can we assume that the new build number is higher and compare `buildNumber > lastWrongBuildNumber8Series` ?

gnufied replied:
Changed it to 8.0.1 and build number `22088125` from - https://kb.vmware.com/s/article/2143832 . AFAIK, that is the last known release. 

My primary concern here is - if vmware will release next version as `8.0.1` or `8.0.2`. If they release it as `8.0.1` and just bump the build numbers, we will be fine.

But if they release next version as `8.0.2`, then last known build number we add here will be meaningless and might cause check to fail. But this also depends on the fact that if `buildNumber` is forever increasing and doesn't resets back when version itself is bumped.  So far it looks like - `buildNumber` is indeed forever increasing, so we might be fine..

jsafrane replied:
The PR looks good to me as it is, with `> minBuildNumber8Series`. If that's consensus, I can lgtm.

jsafrane commented on pkg/operator/vspherecontroller/vspherecontroller.go:
```diff
@@ -494,6 +533,45 @@ func (c *VSphereController) updateConditions(
 	return nil
 }
 
+func (c *VSphereController) addRequiresAdminAck(ctx context.Context, lastCheckResult checks.ClusterCheckResult, name string) error {
```
`name` is not used

jsafrane commented on pkg/operator/vspherecontroller/vspherecontroller.go:
```diff
@@ -494,6 +533,45 @@ func (c *VSphereController) updateConditions(
+	return err
+}
+
+func (c *VSphereController) removeAdminAck(ctx context.Context, name string) error {
```
`name` is not used

jsafrane commented on pkg/operator/vspherecontroller/vspherecontroller.go:
```diff
@@ -494,6 +533,45 @@ func (c *VSphereController) updateConditions(
+		return fmt.Errorf("failed to get admin-gate configmap: %v", err)
+	}
+
+	klog.V(2).Infof("removing admin-gates that is required for CSI migration")
```
I would log it below, when the key exists and is actually removed. Esp. at level 2.

gnufied replied:
moved.

bertinatto commented:
/test e2e-vsphere

jsafrane commented:
/label backport-risk-assessed

jsafrane commented:
/lgtm

dobsonj commented:
/lgtm

sdodson commented:
/retest-required

Phaow commented:
/retest-required

duanwei33 commented:
Doing the pre-merge testing on vsphere 7.0:
The vcenter version is: 7.0.3.0-19717403 and esxi version is 7.0.3.0-19482537, I understand it **doesn't** match the minPatchedVersion7Series and minBuildNumber7Series:
```
	minPatchedVersion7Series = "7.0.3"
	minBuildNumber7Series    = 21424296
```
In operator logs it shows:
```
2023-09-25T03:56:21.812856238Z I0925 03:56:21.812778       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 7.0.3.0-19482537
2023-09-25T03:56:21.812923859Z I0925 03:56:21.812883       1 check_patched_vcenter.go:29] checking for patched version of vSphere for CSI migration: 7.0.3.0-19717403
2023-09-25T03:56:21.812940900Z I0925 03:56:21.812923       1 vsphere_environment_checker.go:91] Overall check result: UpgradeRequiresAdminAck: found version of vCenter which has bugs related to CSI migration. Minimum required, version: 7.0.3, build: 21424296
```
In Overall check result, it shows UpgradeRequiresAdminAck which makes sense to me, but in CSO, it reports:
```
  - lastTransitionTime: "2023-09-25T03:56:36Z"
    message: 'VSphereCSIDriverOperatorCRUpgradeable: VMwareVSphereControllerUpgradeable:
      found version of vCenter which has bugs related to CSI migration. Minimum required,
      version: 7.0.3, build: 21424296'
    reason: AsExpected
    status: "True"
    type: Upgradeable
```
I understand the "Upgradeable" should be "False" because of the min version required.

Also There is no admin-gates generated/added.
  data:
    ack-4.13-kube-1.27-api-removals-in-4.14: Kubernetes 1.27 and therefore OpenShift
      4.14 remove several APIs which require admin consideration. Please see the knowledge
      article https://access.redhat.com/articles/6958395 for details and instructions.

duanwei33 commented:
In the test above, the openshift/cluster-storage-operator#396 is also included.

The cmd to launch build the OCP image.
`build 4.13.0-0.ci,openshift/vmware-vsphere-csi-driver-operator#169,openshift/cluster-storage-operator#396`

duanwei33 commented:
Another test scenario to upgrade 4.12 to 4.13 (Then I have a in-tree pv attached):
1. create 4.12 cluster on vsphere 7.0 with vcenter version 7.0.3.0-19717403 and esxi version 7.0.3.0-19482537
2. Create some pvc with in-tree and dep, make sure the node attached the in-tree volume like:
```
  volumesAttached:
  - devicePath: /dev/disk/by-id/wwn-0x6000c2903b7ea6cf1f1a3b90dfa2e31b
    name: kubernetes.io/vsphere-volume/[host2-local] kubevols/ocp-storage-bhvcg-dyna-pvc-5b442035-69dd-45d6-add0-07aa1f66bc34.vmdk
  - devicePath: /dev/disk/by-id/wwn-0x6000c29758b99b1889af481596b3c3de
    name: kubernetes.io/vsphere-volume/[host2-local] kubevols/ocp-storage-bhvcg-dyna-pvc-f3a19827-a7af-4fd3-8dad-149f1236f7e8.vmdk
  volumesInUse:
  - kubernetes.io/vsphere-volume/[host2-local] kubevols/ocp-storage-bhvcg-dyna-pvc-5b442035-69dd-45d6-add0-07aa1f66bc34.vmdk
  - kubernetes.io/vsphere-volume/[host2-local] kubevols/ocp-storage-bhvcg-dyna-pvc-f3a19827-a7af-4fd3-8dad-149f1236f7e8.vmdk

``` 
3. Upgrade cluster to 4.13, and the CSIMigration is not enabled by default in 4.13 vSphere
4. Check the operator log, seems good
```
I0925 09:04:38.636731       1 vspherecontroller.go:542] Updating admin-gates to require admin-ack for vSphere CSI migration
W0925 09:04:38.807297       1 vspherecontroller.go:522] found version of vCenter which has bugs related to CSI migration. Minimum required, version: 7.0.3, build: 21424296
I0925 09:04:38.810086       1 event.go:285] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-cluster-csi-drivers", Name:"vmware-vsphere-csi-driver-operator", UID:"5607155f-4b2a-42fd-b5af-4d0e0fa4157e", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'ConfigMapUpdated' Updated ConfigMap/admin-gates -n openshift-config-managed:
cause by changes in data.ack-4.13-kube-127-vsphere-migration-in-4.14
```
5. Check the admin-gates, seems good
```
$ oc get ConfigMap/admin-gates -n openshift-config-managed -o yaml
apiVersion: v1
data:
  ......
  ack-4.13-kube-127-vsphere-migration-in-4.14: vSphere CSI migration will be enabled
    in Openshift-4.14. Your cluster appears to be using in-tree vSphere volumes and
    is on a vSphere version that has CSI migration related bugs. See - https://access.redhat.com/node/7011683
    for more information, before upgrading to 4.14.
```
6. Check the storage CO, should the "Upgradeable" mark as "False"? 
```
  - lastTransitionTime: "2023-09-25T07:20:13Z"
    message: 'VSphereCSIDriverOperatorCRUpgradeable: VMwareVSphereControllerUpgradeable:
      found version of vCenter which has bugs related to CSI migration. Minimum required,
      version: 7.0.3, build: 21424296'
    reason: AsExpected
    status: "True"
    type: Upgradeable
```
7. check the output of "oc adm upgrade", it doesn't contains the storage reason, same concern as #6
```
$ oc adm upgrade
Cluster version is 4.13.0-0.test-2023-09-25-024658-ci-ln-s7nwywt-latest

Upgradeable=False

  Reason: MultipleReasons
  Message: Cluster should not be upgraded between minor versions for multiple reasons: AdminAckRequired,ResourceDeletesInProgress
  * Kubernetes 1.27 and therefore OpenShift 4.14 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958395 for details and instructions. vSphere CSI migration will be enabled in Openshift-4.14. Your cluster appears to be using in-tree vSphere volumes and is on a vSphere version that has CSI migration related bugs. See - https://access.redhat.com/node/7011683 for more information, before upgrading to 4.14.
  * Cluster minor level upgrades are not allowed while resource deletions are in progress; resources=clusterrolebinding "default-account-openshift-machine-config-operator"
```

duanwei33 commented:
After check with @gnufied  offline, so only both `unsupported version` and `in-tree volume present` satisfied, the ConfigMap/admin-gates added and we could see the storage reason in "oc adm upgrade" output. And we **don't** add the **upgradable: false** in storage CO.
So the only comment is:

- In vSphere 8 env, we need to compare with vSphere 8 condition, and it should **>** "8.0.1.0-22088125"
- currently storage reason is in the same line with "remove several APIs" in "oc adm upgrade" output, which is not easy to catch.
```
$ oc adm upgrade
Cluster version is 4.13.0-0.test-2023-09-25-024658-ci-ln-s7nwywt-latest

Upgradeable=False

  Reason: MultipleReasons
  Message: Cluster should not be upgraded between minor versions for multiple reasons: AdminAckRequired,ResourceDeletesInProgress
  * Kubernetes 1.27 and therefore OpenShift 4.14 remove several APIs which require admin consideration. Please see the knowledge article https://access.redhat.com/articles/6958395 for details and instructions. vSphere CSI migration will be enabled in Openshift-4.14. Your cluster appears to be using in-tree vSphere volumes and is on a vSphere version that has CSI migration related bugs. See - https://access.redhat.com/node/7011683 for more information, before upgrading to 4.14.
  * Cluster minor level upgrades are not allowed while resource deletions are in progress; resources=clusterrolebinding "default-account-openshift-machine-config-operator"
```

gnufied commented:
>currently storage reason is in the same line with "remove several APIs" in "oc adm upgrade" output, which is not easy to catch.

As such we don't control how the message is displayed in `oc adm upgrade`. It is concatenated by CVO, but I will ask around and check if there is a way to improve how the message appears.

gnufied commented:
It should also be noted that, as each admin-gate is acked, the message gets updated and if migration warning is last thing remaining, it will be only thing that will get displayed. So regardless of how message appears, customers can't upgrade without acking all pending `admin-gates` checks.

gnufied commented:
> In vSphere 8 env, we need to compare with vSphere 8 condition, and it should > "8.0.1.0-22088125"

Also @duanwei33 we are actually already comparing against `8.0.1.0-22088125`. The problem was logged message when 8 series is detected. It should return warning message from 8 series and not 7 series. 

Anyways, I have addressed all your concerns (except about messages getting concatenated in `oc adm upgrade`):

1. Warning message from proper vSphere series are reported now. 
2. I have removed reporting of `admin-ack` errors in `Upgradeable` condition, because I think this is just confusing to end users. We still log the admin-acks but we no longer report them in conditions.

duanwei33 commented:
In csi-driver operator log:
```
I0926 09:14:41.015903       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 8.0.1.0-22088125
I0926 09:14:41.021258       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 8.0.1.0-22088125
I0926 09:14:41.037725       1 check_nodes.go:174] checking for patched version of ESXi for CSI migration: 8.0.1.0-22088125
I0926 09:14:41.037813       1 check_patched_vcenter.go:29] checking for patched version of vSphere for CSI migration: 8.0.1.0-22088981
I0926 09:14:41.037840       1 vsphere_environment_checker.go:91] Overall check result: UpgradeRequiresAdminAck: host host-1021 is on ESXI version 8.0.1.0-22088125, minimum required for CSI migration: > 8.0.1-22088125
```

There is no message as we do allow such cluster upgrade.
```
  - lastTransitionTime: "2023-09-26T00:14:36Z"
    message: 'VSphereCSIDriverOperatorCRUpgradeable: VMwareVSphereControllerUpgradeable:
      All is well'
    reason: AsExpected
    status: "True"
    type: Upgradeable
```

So the test looks good from my side.

duanwei33 commented:
/label qe-approved
/label cherry-pick-approved

jsafrane commented:
/lgtm
